src/pool/srv_pool.c

```diff
@@ -922,6 +925,63 @@ pool_svc_alloc_cb(d_iov_t *id, struct ds_rsvc **rsvc)
 	return rc;
 }
 
+struct ds_pool_evict_arg {
+	struct pool_svc *svc;
+	d_rank_t	rank;
+};
+
+void
+ds_pool_evict_rank_ult(void *data)
+{
+	struct ds_pool_evict_arg *arg = data;
+	int			 rc;
+
+	rc = ds_pool_evict_rank(arg->svc->ps_uuid, arg->rank);
+
+	D_DEBUG(DB_MGMT, DF_UUID" evict rank %u : rc %d\n",
+		DP_UUID(arg->svc->ps_uuid), arg->rank, rc);
+
+	D_FREE_PTR(arg);
+}
+
+static void
+ds_pool_crt_event_cb(d_rank_t rank, enum crt_event_source src,
+		     enum crt_event_type type, void *arg)
+{
+	struct ds_pool_evict_arg	*ult_arg;
+	struct pool_svc			*svc = arg;
+	int				rc = 0;
+
+	/* Only used for evict the rank for the moment */
+	if (src != CRT_EVS_SWIM || type != CRT_EVT_DEAD) {
+		D_DEBUG(DB_MGMT, "ignore src/type %u/%u\n", src, type);
+		return;
+	}
+
+	if (svc->ps_disable_evict) {
+		D_DEBUG(DB_MGMT, "event response is disabled\n");
+		return;
+	}
+
+	D_ALLOC_PTR(ult_arg);
+	if (ult_arg == NULL) {
+		D_ERROR("allocation failed\n");
+		return;
+	}
+
+	/* Since this is called from swim_progress, let's create another
+	 * ult to do the job asynchronously, so this can return quickly.
+	 */
+	ult_arg->svc = svc;
+	ult_arg->rank = rank;
+	rc = dss_ult_create(ds_pool_evict_rank_ult, ult_arg, DSS_ULT_MISC,
+			    DSS_TGT_SELF, 0, NULL);
```

Comment:
if there are multiple events reported at the same time, seems ds_pool_evict_rank_ult cannot ensure to be executed by the order? may need to insert the event to a list, and in the ult consume the list to ensure the order.

Reply:
Ok, seems not must to keep that order...  
Another problem is that if the server crash after get the swim event and before ds_update_pool by ULT, will it cause the event ignored (even after new srv leader take over)?

Reply:
oh, there is a collective operation in ds_pool_tgt_map_update(), which will yield and wait. it has to be an ULT. 
Hmm, for evict, it does not have to be in order. probably later.

Reply:
agree that need not to be in order, but seems that current code cannot ensure the event cannot be lost/ignored.
maybe need to store it in RDB when get the node failure event from cart, and the ULT read the event from RDB and handle it, then when this server failed and another one step up it still can pickup from RDM and handle it. But can be done in following patch.

Reply:
I really don't like creating orphan ULTs (i.e., those that not explicitly joined and freed) for core ULTs (i.e., not for incoming RPCs). They may still exist when the pool service leader steps down or when the pool is destroyed. Using pool_svc ref and pool_svc leader ref may mitigate the issue, but it would be hard to know who are holding the references we are waiting for. I believe the queue and queue consumer ULT model I used in my prototype is better in the long run.

My thought for event losses was that at step up time we need to compare the pool map with the SWIM rank status and requeue the ranks that are IN in pool map but DEAD in SWIM. Then the queue consumer ULT can process the queue as usual.

Reply:
> I really don't like creating orphan ULTs (i.e., those that not explicitly joined and freed) for core ULTs (i.e., not for incoming RPCs). They may still exist when the pool service leader steps down or when the pool is destroyed. Using pool_svc ref and pool_svc leader ref may mitigate the issue, but it would be hard to know who are holding the references we are waiting for. I believe the queue and queue consumer ULT model I used in my prototype is better in the long run.

Hmm, initially I had thought to update the pool map synchronously. but got some issues since update pool map might be "blocked" due to collective operation inside ds_pool_tgt_map_update(). So I created an ULT here to finish the pool map update.

As for whether we should invite a specific mechanism to manage this ULT, instead of trusting the scheduler. I prefer the "simpler" solution.  Though I did not see your prototype yet, sorry about that.   Is this consumer ULT model only for this swim event? Are there any other use cases here?  

> 
> My thought for event losses was that at step up time we need to compare the pool map with the SWIM rank status and requeue the ranks that are IN in pool map but DEAD in SWIM. Then the queue consumer ULT can process the queue as usual.

Reply:
Essentially, the event callback could simply append events to a queue in pool_svc. A daemon ULT, started at step up time and stopped at step down time, consumes events from the queue. Regardless of how rapidly SWIM generates events, there would be only one such ULT. This ULT would be the single place where we make exclusion decisions based on criteria: some event timestamps, whether data loss might happen, whether auto exclusion has been disabled, etc. Like in an I/O scheduler, the queue could allow us to merge events or handle a batch of them in one RDB TX. Conceptually it would be simpler and there'd be no orphan ULTs.

This is a general pattern used for rdb's RPC reply handling and step up/down event handling too.

Reply:
Ok, thanks for the explanation.  But I am not sure these are for this ticket. I was told to only adding a callback in DAOS to handle the swim DEAD event.  This seems much bigger than what it looks like.  If you have patch already, I do not mind closing this one. Otherwise, can we do this later? since this seems target for 0.9?

Reply:
OK, I didn't know there's pressure to get this landed. Could you make ds_rsvc_get external and take a reference for ult_arg.svc, and return the reference once the ULT is done with ult_arg.svc, please?

Reply:
Hmm, I thought about that. but this svc will be registered during step up, and unregistered during step down. So step_up/down process make sure the svc is always hold? no?

Reply:
Because these ULTs are orphans, we may have:

    step up
    register event cb
    create ULT x for event x
    unregister event cb
    step down
    ULT x still running

Reply:
ah, you are right.

Reply:
For this particular case, I think we can just eliminate the collective calls in ds_pool_tgt_map_update() and call it directly. What do you think?

Reply:
Hmm, I do not see creating an ULT cause any trouble (or complication) here.  Besides, it needs to update child map on each xstream as normal pool map update anyway. Or I miss sth here?

Reply:
I don't see why we dup a uint32_t read-only pool map version for each xstream in first place, can't we just read the pool map version in per-node ds_pool?

Reply:
I assume we do not accessing per node structure per xstream. Besides, it also update the pool map over client stack. 

Btw: I really do not see much problem to create an extra ULT here. what is your concern here?   This is from cart progress, and we need return asap anyway.

Reply:
Well, I just think that if it's not necessary, we can simplify it to avoid the troubles mentioned by Liwei. I still think that all xstreams reading the same uint32 pool map version should be fine, but I do agree that calling client stack directly here could be dangerous.

I was just wondering if that part (ds_rebuild_pool_map_update()) could be moved from this comment pool map update function to somewhere else, of course, that's out of scope of this patch.

Reply:
Hmm, I thought Li wei's concern is orphan ULT, instead of not creating ULT.

Reply:
Right, I would also create ULTs (though only one). The RDB operations and pool metadata lock operations are also blocking. Thanks for taking svc refs.